src/middleware/externalRequests.js

```diff
@@ -63,8 +77,253 @@ const prepareRequestConfig = requestDetails => {
+    ctx.body[request.id] = body
+  }
+}
+
+const createAxiosConfig = (request, body) => {
+  const axiosConfig = {
+    url: request.url,
+    method: request.method,
+    body: body
+  }
+
+  if (
+    request.credentials &&
+    request.credentials.username &&
+    request.credentials.password
+  ) {
+    axiosConfig.auth = {
+      username: request.credentials.username,
+      password: request.credentials.password
+    }
+  }
+
+  if (request.headers) {
+    axiosConfig.headers = request.headers
+  }
+
+  return axiosConfig
+}
+
+const sendMappedObject = (ctx, axiosConfig, request, body) => {
+  const reqTimestamp = new Date()
+  let response, error, responseTimestamp
+
+  return axios(axiosConfig)
+    .then(resp => {
+      response = resp
+      response.body = resp.data
+      responseTimestamp = new Date()
+
+      if (request.primary) {
+        setKoaResponseBodyFromPrimary(ctx, request, response.body)
+
+        ctx.status = response.status
+      } else {
+        setKoaResponseBodyFromPrimary(ctx, request, response.body)
+      }
+    })
+    .catch(err => {
+      responseTimestamp = new Date()
+
+      const result = handleRequestError(ctx, request, err)
+      response = result.response
+      error = result.error
+    })
+    .finally(() => {
+      if (ctx.request.header && ctx.request.header['X-OpenHIM-TransactionID']) {
+        const orch = createOrchestration(
+          request,
+          body,
+          response,
+          reqTimestamp,
+          responseTimestamp,
+          request.id,
+          error
+        )
+
+        ctx.orchestrations.push(orch)
+      }
+    })
+}
+
+const constructOpenhimResponse = (ctx, responseTimestamp) => {
+  const response = ctx.response
+  const orchestrations = ctx.orchestrations
+  const statusText = ctx.statusText
+  const respObject = {}
+
+  if (response) {
+    if (response.headers) {
+      respObject.headers = response.headers
+    }
+    if (response.status) {
+      respObject.status = response.status
+    }
+    if (response.body) {
+      respObject.body = response.body
+    }
+    if (response.timestamp) {
+      respObject.timestamp = response.timestamp
+    } else if (responseTimestamp) {
+      respObject.timestamp = responseTimestamp
+    }
+  }
+
+  if (readError) {
+    mediatorConfigJson = {
+      urn: 'undefined'
+    }
+  }
+
+  const body = {
+    'x-mediator-urn': mediatorConfigJson.urn,
+    status: statusText,
+    response: respObject,
+    orchestrations: orchestrations
+  }
+
+  ctx.body = JSON.stringify(body)
+}
+
+const setStatusText = ctx => {
+  if (ctx.primaryReqFailError) {
+    ctx.statusText = 'Failed'
+  } else if (!ctx.primaryReqFailError && ctx.secondaryFailError) {
+    ctx.statusText = 'Completed with error(s)'
+  } else if (
+    !ctx.primaryReqFailError &&
+    !ctx.secondaryFailError &&
+    (ctx.primaryCompleted || ctx.secondaryCompleted)
+  ) {
+    ctx.statusText = 'Completed'
+  } else {
+    ctx.statusText = 'Successful'
+  }
```

Comment:
@bradsawadye this all seems very specific to the OpenHIM. Could it not be extracted as well.

Reply:
done in https://github.com/jembi/openhim-mediator-mapping/pull/36/commits/e671af21b1ef83c3828dbd0adf3e044160ca9074